### DISCUSSION TITLE
Language switcher pop-up menu doesn't connect to taskbar on top.

### DISCUSSION THREAD

**Discussion by ghost** [valinet/ExplorerPatcher#629] [Q&A] [answered]
First of all, I want to to say big Thank You for this patcher, makes Win 11 usable for me.

The only tiny problem I have is visual. I use taskbar on top (win 10 style), and when I mouse click on language switcher (win 11 style) the pop-up menu opens at the bottom, somewhat cropped and disconnected from taskbar at top. If I try to use lang. switcher with win 10 style, then mouse click does nothing at all.

So my question is: Is there any way to "glue" this pop-up menu to taskbar on top? Both win11 / win10 style will work for me.

Thanks.

<img width="305" alt="Untitled" src="https://user-images.githubusercontent.com/97186640/148263406-f757ac19-8600-4cd6-a650-cd2b8c02db1e.png">

**Comment by valinet** [comment, marked as answer]
Hi

> First of all, I want to to say big Thank You for this patcher, makes Win 11 usable for me.

You're welcome, I am glad that it works for you.

> The only tiny problem I have is visual. I use taskbar on top (win 10 style), and when I mouse click on language switcher (win 11 style) the pop-up menu opens at the bottom, somewhat cropped and disconnected from taskbar at top. If I try to use lang. switcher with win 10 style, then mouse click does nothing at all.

Indeed, not pleasant. The Windows 10 language switcher actually shows, but it still shows above the taskbar, which is already on the top. Apparently, the coordinates where it shows are never adjusted. Fortunately, I had some inspiration today, looked a bit on `InputSwitch.dll`'s disassembly (the component that hosts the language switcher), and was able to figure out what is invoked by `explorer.exe` when the menu switcher is shown (`CInputSwitchControl::ShowInputSwitch`) and one of the parameters is a rectangle containing the coordinates of the tray indicator for the switcher, apparently. The position of the switcher is computed based on that, always as if the taskbar is on the bottom. With some "Maths magic" I quickly put together some code that adjusts that rectangle in order to correct the position where it shows and have it display properly, wherever the taskbar is placed. It seems to work. Also, while I was at it, I took the time to clear up and make the language switcher list more clear, without those cryptic "LOGONUI", "UAC", "SETTINGPAGE" and "OOBE" options). There are now 3 options:

* Windows 11 (default)
* Windows 10 (with link to "Language Preferences")
* Windows 10

tl;dr I fixed this for the Windows 10 language switcher. I don't know how to fix the Windows 11 language switcher, but hopefully the work I did today gets the job done, for the time being.

![image](https://user-images.githubusercontent.com/6503598/148297279-5868e63a-d11d-46f3-9e87-a848bbe5a7bb.png)

The changes will make their way into a new pre-release, 41.2, very soon.

Thank you
